akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -3,70 +3,464 @@
-      (new SinkholeSubscriber(Promise()), Future.successful(context.effectiveAttributes))
+  class AttributesSink(_initialAttributes: Attributes = Attributes.none) extends GraphStageWithMaterializedValue[SinkShape[Any], Attributes] {
+
+    val in = Inlet[Any]("in")
 
-    override protected def newInstance(shape: SinkShape[Nothing]): SinkModule[Nothing, Future[Attributes]] =
-      new AttributesSink(attributes, shape)
+    override protected def initialAttributes: Attributes = _initialAttributes
+    override val shape = SinkShape(in)
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Attributes) = {
+      val logic = new GraphStageLogic(shape) {
+        override def preStart(): Unit = {
+          pull(in)
+        }
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            grab(in)
+            pull(in)
+          }
+        })
+      }
 
-    override def withAttributes(attr: Attributes): SinkModule[Nothing, Future[Attributes]] =
-      new AttributesSink(attr, amendShape(attr))
+      (logic, inheritedAttributes)
+    }
   }
 
+  class ThreadNameSnitchingStage(initialDispatcher: String) extends GraphStage[SourceShape[String]] {
+    val out = Outlet[String]("out")
+    override val shape = SourceShape.of(out)
+    override protected def initialAttributes: Attributes = ActorAttributes.dispatcher(initialDispatcher)
+    def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+      setHandler(out, new OutHandler {
+        def onPull(): Unit = {
+          push(out, Thread.currentThread.getName)
+          completeStage()
+        }
+      })
+    }
+
+  }
+
+  def whateverAttribute(label: String): Attributes = Attributes(WhateverAttribute(label))
+  case class WhateverAttribute(label: String) extends Attribute
 }
 
-class AttributesSpec extends StreamSpec {
+class AttributesSpec extends StreamSpec(ConfigFactory.parseString(
+  """
+    my-dispatcher {
+      type = Dispatcher
+      executor = "thread-pool-executor"
+      thread-pool-executor {
+        fixed-pool-size = 1
+      }
+      throughput = 1
+    }
+  """).withFallback(Utils.UnboundedMailboxConfig)) {
+
   import AttributesSpec._
 
   val settings = ActorMaterializerSettings(system)
     .withInputBuffer(initialSize = 2, maxSize = 16)
 
   implicit val materializer = ActorMaterializer(settings)
 
-  "attributes" must {
+  "an attributes instance" must {
+
+    val attributes = Attributes.name("a") and Attributes.name("b") and Attributes.inputBuffer(1, 2)
+
+    "give access to the least specific attribute" in {
+      attributes.leastSpecific[Name] should ===(Some(Attributes.Name("a")))
+    }
+
+    "give access to the most specific attribute value" in {
+      attributes.mostSpecific[Name] should ===(Some(Attributes.Name("b")))
+    }
+
+    "prefer most specific important attribute to most specific attribute" in {
```

Comment:
Test description got it wrong, following your naming I guess it's the "least specific + important" one that wins, since those behave like overrides the outer-most seem to win

Reply:
Nope. Most specific marked as important first, if there is no attribute of the type marked important fall back to most specific "normal"

Reply:
Hmm yeah I see... Perhaps good enough, though a bit weird actually. `!important` semantics from css allow overriding an important one AFAIR, so this is more like `.lock()` I guess.
Perhaps it's good enough for our case though hm...

Reply:
The idea was to mimic the CSS semantics indeed, and there the most specific !important wins (I think in CSS the order of the !important blocks in the CSS files also plays a role which we certainly should not try to mimic).

Reply:
(For reference from: https://www.w3.org/TR/CSS22/cascade.html#cascading-order and https://www.w3.org/TR/CSS22/cascade.html#important-rules)